com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/Nodes/HDSampleBufferNode.cs

```diff
@@ -0,0 +1,230 @@
+                    channelCount = 2;
+                    break;
+                case BufferType.PostProcessInput:
+                case BufferType.BlitSource:
+                    AddSlot(new ColorRGBAMaterialSlot(k_OutputSlotId, k_OutputSlotName, k_OutputSlotName, SlotType.Output, Color.black, ShaderStageCapability.Fragment));
+                    channelCount = 4;
+                    break;
+            }
+
+            RemoveSlotsNameNotMatching(new[]
+            {
+                k_ScreenPositionSlotId,
+                k_SamplerInputSlotId,
+                k_OutputSlotId,
+            });
+        }
+
+        public override void CollectShaderProperties(PropertyCollector properties, GenerationMode generationMode)
+        {
+            if (generationMode.IsPreview())
+                return;
+
+            if (bufferType == BufferType.BlitSource)
+            {
+                properties.AddShaderProperty(new Texture2DArrayShaderProperty
+                {
+                    overrideReferenceName = "_MainTex",
+                    displayName = "_MainTex",
+                    hidden = true,
+                    generatePropertyBlock = true,
+                    isMainTexture = true,
+                });
+            }
+            else if (bufferType == BufferType.PostProcessInput)
+            {
+                properties.AddShaderProperty(new Texture2DArrayShaderProperty
+                {
+                    overrideReferenceName = nameof(HDShaderIDs._CustomPostProcessInput),
+                    displayName = nameof(HDShaderIDs._CustomPostProcessInput),
+                    hidden = true,
+                    generatePropertyBlock = true,
+                    isMainTexture = true,
+                });
+            }
+        }
+
+        string GetFunctionName() => $"Unity_HDRP_SampleBuffer_{bufferType}_$precision";
+
+        public void GenerateNodeFunction(FunctionRegistry registry, GenerationMode generationMode)
+        {
+            // Preview SG doesn't have access to HDRP depth buffer
+            if (!generationMode.IsPreview())
+            {
+                registry.ProvideFunction(GetFunctionName(), s =>
+                {
+                    s.AppendLine("#include \"Packages/com.unity.render-pipelines.high-definition/Runtime/Material/NormalBuffer.hlsl\"");
+                    s.AppendLine("#include \"Packages/com.unity.render-pipelines.high-definition/Runtime/Material/Builtin/BuiltinData.hlsl\"");
+
+                    s.AppendLine("$precision{1} {0}($precision2 uv, SamplerState samplerState)", GetFunctionName(), channelCount);
+                    using (s.BlockScope())
+                    {
+                        switch (bufferType)
+                        {
+                            case BufferType.NormalWorldSpace:
+                                s.AppendLine("uint2 pixelCoords = uint2(uv * _ScreenSize.xy);");
```

Comment:
Consider wrapping this into simpler functions

Reply:
In C# or the generated code?

Reply:
I was hoping we could have something like a straight hlsl files we could include here, keep this complex code in such files, and just call the functions from here. If its too hard to do dont worry about it.